### PR TITLE
feature(store): make list_* methods async generators

### PR DESCRIPTION
**Pull request by jhamman**
This goes on top of https://github.com/zarr-developers/zarr-python/pull/1743
TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: src/zarr/v3/store/memory.py [hunk 1]
```diff
@@ -67,20 +68,18 @@ async def delete(self, key: str) -> None:
...
+            yield key
+
+    async def list_prefix(self, prefix: str) -> AsyncGenerator[str, None]:
+        for key in self._store_dict:
+            if key.startswith(prefix):
+                yield key
+
+    async def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
+        print('prefix', prefix)
+        print('keys in list_dir', list(self._store_dict))
```
Comment (jhamman) [inline src/zarr/v3/store/memory.py]:
```suggestion
```

File: src/zarr/v3/group.py [hunk 2]
```diff
@@ -302,22 +303,26 @@ async def members(self) -> AsyncGenerator[tuple[str, AsyncArray | AsyncGroup], N
...
+                    # TODO: performance optimization -- load children concurrently   
+                    child = await self.getitem(key)
+                    # keyerror is raised when `key``names an object in the store
+                    # in which case `subkey` cannot be the name of a sub-array or sub-group.
+                    yield key, child
+                except KeyError:
+                    # keyerror is raised when `subkey` names an object (in the object storage sense),
+                    # as opposed to a prefix, in the store under the prefix associated with this group
+                    # in which case `subkey` cannot be the name of a sub-array or sub-group.
+                    logger.warning(
```
Comment (d-v-b) [inline src/zarr/v3/group.py]: Are both of these comments necessary, or can we just use the second one (after changing `subkey` to `key`)?

Comment (d-v-b) [reply src/zarr/v3/group.py]:
```suggestion
                    yield key, child
                except KeyError:
                    # keyerror is raised when `key` names an object (in the object storage sense),
                    # as opposed to a prefix, in the store under the prefix associated with this group.
                    # In this case, `key` cannot be the name of a sub-array or sub-group.
                    logger.warning(
```

File: src/zarr/v3/store/memory.py [hunk 3]
```diff
@@ -67,20 +68,16 @@ async def delete(self, key: str) -> None:
...
-                {
-                    key.removeprefix(prefix + "/").split("/")[0]
-                    for key in self._store_dict
-                    if (key.startswith(prefix + "/") and key != prefix)
-                }
-            )
+    async def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
+        for key in self._store_dict:
+            if key.startswith(prefix + "/") and key != prefix:
+                yield key.strip(prefix + "/").rsplit("/", maxsplit=1)[0]
```
Comment (d-v-b) [inline src/zarr/v3/store/memory.py]: Confusingly, `key.strip("foo")` treats "foo" as a set, e.g. `{"f", "o"}`: 

```python
>>> 'ba/x/bbaa'.strip('ab')
'/x/'
``` 

I think `removeprefix` does what we want here:

```python
>>> 'ba/x/bbaa'.removeprefix('ab')
'ba/x/bbaa'
>>> 'ba/x/bbaa'.removeprefix('ba')
'/x/bbaa'
```

```suggestion
                yield key.removeprefix(prefix + "/").rsplit("/", maxsplit=1)[0]
```

**Review by d-v-b — APPROVED** [review]

**Pull request merged**